chore(suite): move from NextJS to custom build pipeline

keraf opened the pull request:
This is quite a big change in how Suite works under the hood. We mainly relied on NextJS providing all the bells and whistles, but this change introduces our own build/development pipeline with dependencies that we choose. 

### Why switch from NextJS?
Given the type of project Suite is, NextJS had only few benefits that came with some negative costs.
- We don't use SSR and SSG creates more overhead. In a web application like ours, CSR is the way to go.
- Rebuilding would crash at best and freeze the whole machine at worst, this was happening way to frequently.
- We have little to no control over the build pipeline. Sure we can add plugins and tweak the webpack configuration in NextJS, but it's far from ideal to reach certain performance goals.

### Benefits of this change
- Builds are faster.
- Latest libraries are being used (for example React Fast Refresh). We can take advantage of new features/methods as well as patch ourselves for vulnerabilities sooner.
- During development, all routes/views are generated right away, making the navigation quicker.
- Doesn't crash or freeze the whole machine when building.
- More control over how we bundle the project (smaller chunks, can take advantage of parallelization of HTTP2).
- Builds are lighter to load (lazy loading of routes & heavy dependencies).
- Browser detect is now a React component and lazy loads the rest of the app. If Suite web is not compatible with the user browser, less than 500kb will be loaded.
- Assets are automatically optimized at build time. We can use original assets in the repo.

### And the downsides
- Navigating between routes on web will display a loader. This is because routes are lazy loaded.

### Future
Having control over our own build pipelines opens a lot of doors for future improvements. Adopting new libraries & tools, or changing them, will be easier as we are not restricted by NextJS as a dependency. We can for example imagine migrating to Snowpack in the near future.

### Tasks
- [x] Suite web with custom pipeline
- [x] Suite desktop with custom pipeline
- [x] CI working
- [x] Tests working
- [x] Development features
  - [x] Aliases
  - [x] Hot Module Reloading (now using Fast Refresh)
  - [x] Source maps
- [x] Working router
- [x] Headers (using Helmet)
- [x] Faster build/rebuilds for dev & prod
- [x] Code splitting 
- [x] Generate HTML files for routes
- [x] Integrate Browser Detect into the pipeline
- [x] Documentation

goodhoko commented:
Wanted to give this a quick review but was scared away by that huge diff all within a single commit. Do you plan to break this down to smaller, atomic, more digestable commits? That would really help to anyone wanting to review this. .)

goodhoko commented:
@keraf I pulled this branch and ran `yarn suite:dev`, `yarn suite:dev:desktop` and `yarn suite:build:web` with no problems at all! The app just seems to work. 👍

keraf commented:
@trezor/qa This one is going to be a bitch to test. Basically everything needs tested on web and desktop (only one OS, up for you to choose) :sweat_smile: The main focus should be on differences with the current develop builds as well as issues that don't exist in develop. I guess it should mainly be focused on the big pieces (sending, receiving, viewing/searching transactions, playing with settings, etc...).

To give a bit more background to this change, it's changing the way the project is built but the end result is expected to be the same to the user (with some speed / memory / page weight improvements). One notable change on web is loaders between pages. This has been done as new code is being lazy loaded when the user navigates. The rest should be the same as before.

If you need more info or have questions, feel free to DM me.

szymonlesisz commented on packages/suite/src/views/dashboard/components/PortfolioCard/components/DashboardGraph/index.tsx:
```diff
@@ -6,9 +6,7 @@ import styled from 'styled-components';
-// https://github.com/zeit/next.js/issues/4768
-// eslint-disable-next-line import/no-webpack-loader-syntax
-import GraphWorker from 'worker-loader?filename=static/[hash].worker.js!../../../../../../workers/graph.worker';
+import GraphWorker from '../../../../../../workers/graph.worker';
```
```suggestion
import GraphWorker from '@suite/workers/graph.worker';
```
alias should work here, no?

keraf replied:
cf1975a

szymonlesisz commented on packages/suite/src/utils/wallet/exportTransactions.ts:
```diff
@@ -117,7 +125,7 @@ const preparePdf = (
     coin: Network['symbol'],
     accountName: string,
-): TDocumentDefinitions => {
+): any => {
```
forgotten any?

keraf replied:
cf1975a

szymonlesisz commented on packages/suite/src/utils/suite/nextjs.ts:
```diff
@@ -1,10 +1 @@
-
-    return staticPath;
-};
+export const resolveStaticPath = (path: string) => `${process.env.ASSET_PREFIX}/static/${path}`;
```
should it be called `nextjs` util?

keraf replied:
cf1975a

szymonlesisz commented on package.json:
```diff
@@ -70,8 +71,13 @@
+        "ts-node": "^9.1.1",
+        "tsconfig-paths": "^3.9.0",
+        "typescript": "4.1.3",
+        "webpack": "^5.28.0",
```
webpack dependency in root?
possible collision with other/older webpacks (blockchain-link, nextjs landing page, suite-data...)

keraf replied:
Landing page and blockchain-link are fine (just tested), suite-data won't be needing webpack anymore.

szymonlesisz commented on shell.nix:
```diff
@@ -1,8 +1,8 @@
   (builtins.fetchTarball {
-    url = "https://github.com/NixOS/nixpkgs/archive/0b8799ecaaf0dc6b4c11583a3c96ca5b40fcfdfb.tar.gz";
-    sha256 = "11m4aig6cv0zi3gbq2xn9by29cfvnsxgzf9qsvz67qr0yq29ryyz";
+    url = "https://github.com/NixOS/nixpkgs/archive/f8929dce13e7.tar.gz";
```
is this change necessary? i was able to run it with previous (current develop) derivation without error

keraf replied:
cf1975a

szymonlesisz commented on packages/suite/src/utils/wallet/exportTransactions.ts:
```diff
@@ -1,11 +1,8 @@
 import { trezorLogo } from '@suite-constants/b64images';
 
-pdfMake.vfs = pdfFonts.pdfMake.vfs;
+import type { TDocumentDefinitions } from 'pdfmake/interfaces';
```
import `type` convention. i know it's possible but we are not using it anywhere.
should we?

szymonlesisz commented on packages/suite/src/support/suite/Router.tsx:
```diff
@@ -1,47 +1,31 @@
+        const onPopState = () => {
+            const canGoBack = actions.onBeforePopState();
+            if (!canGoBack) {
+                history.go(1);
```
you can block transition before it happens
https://github.com/ReactTraining/history/blob/master/docs/blocking-transitions.md

keraf replied:
This is based of the old code and works for now. Added this to the follow up issue and will address in a different PR: #3670

szymonlesisz commented on packages/suite/src/components/suite/modals/QrScanner/index.tsx:
```diff
@@ -3,12 +3,10 @@ import * as URLS from '@suite-constants/urls';
 import { Icon, colors, P } from '@trezor/components';
 import { UserContextPayload } from '@suite-actions/modalActions';
-import dynamic from 'next/dynamic';
+import QrReader from 'react-qr-reader';
```
there was dynamic load because of bundle size? same as `zxcvbn` in passwordStrenght ?

keraf replied:
16116f3

szymonlesisz commented on packages/suite-web/src/index.tsx:
```diff
@@ -0,0 +1,14 @@
+
+const Main = lazy(() => import('./Main'));
+const Index = () => (
+    <BrowserDetect>
```
browser detection should be done before whole app is loaded.
first reason: you may have unsupported browser (like some IE) which will fail on js runtime (missing polyfills) then "unsupported page" will never render. you will se a blank page.
second reason: you have to download the whole app (~5MB) to get information  that you are not allowed to use it

keraf replied:
Browser detect has been reverted (c82d5f6).

goodhoko commented on docs/misc/build.md:
```diff
@@ -0,0 +1,21 @@
+The following commands are available in this package (using `yarn run` at the root of the package or `yarn workspace @trezor/suite-build run` at the root of the project):
+| Command | Description |
+| --- | --- |
+| dev:web | Runs a watch build of `suite-web` with development settings and servies it. |
```
```suggestion
| dev:web | Runs a watch build of `suite-web` with development settings and serves it. |
```

keraf replied:
cf1975a

goodhoko commented on docs/misc/build.md:
```diff
@@ -0,0 +1,21 @@
+| Command | Description |
+| --- | --- |
+| dev:web | Runs a watch build of `suite-web` with development settings and servies it. |
+| build:web | Runs a production build of `suite-web`. |
```
```suggestion
| build:web | Builds a production version of `suite-web`. |
```

keraf replied:
cf1975a

goodhoko commented on docs/misc/build.md:
```diff
@@ -0,0 +1,21 @@
+| --- | --- |
+| dev:web | Runs a watch build of `suite-web` with development settings and servies it. |
+| build:web | Runs a production build of `suite-web`. |
+| dev:desktop | Runs a watch build of `suite-desktop` with development settings, servies it and runs the Electron wrapper. |
```
```suggestion
| dev:desktop | Runs a watch build of `suite-desktop` with development settings, serves it and runs the Electron wrapper. |
```

Also what is the electron wrapper? This makes it sound like it's an independent proces of the actual app...?

keraf replied:
cf1975a

goodhoko commented on docs/misc/build.md:
```diff
@@ -0,0 +1,21 @@
+| dev:web | Runs a watch build of `suite-web` with development settings and servies it. |
+| build:web | Runs a production build of `suite-web`. |
+| dev:desktop | Runs a watch build of `suite-desktop` with development settings, servies it and runs the Electron wrapper. |
+| build:desktop | Runs a production build of `suite-desktop`. |
```
```suggestion
| build:desktop | Builds a production version of `suite-desktop`. |
```

keraf replied:
cf1975a

goodhoko commented on docs/misc/build.md:
```diff
@@ -0,0 +1,21 @@
+# Front-End Build
```
I'd be happier if this was placed in `packages/suite-build/README.md` but I'll leave that on you. .)

goodhoko commented on docs/misc/build.md:
```diff
@@ -0,0 +1,21 @@
+The front-end build of Suite is handled by Webpack configurations inside the `suite-build` package.
+
+The folder structure is as follows:
+- configs: Contains the Webpack configuration files. `base.webpack.config.ts` serves as a comon base for all other configurations. The other files in this folder are project specific such as `web.webpack.config.ts` or `desktop.webpack.config.ts` for `suite-web` and `suite-desktop` respectively. 
```
I think it would be nicer for the config files to be placed in their respective packages and only the shared config to be pulled out either to the root of the repo or to this `suite-build` package.

In my ideal world each package is self-contained except dependencies and artefacts inherited from higher level directories and there are no 'side cars'.

goodhoko commented on packages/suite-build/configs/base.webpack.config.ts:
```diff
@@ -0,0 +1,148 @@
+    devtool: isDev ? 'eval-source-map' : false,
+    devServer: {
+        port: 3000,
+        hot: isDev, // HMR
```
What does `HMR` stands for?

keraf replied:
Hot Module Reloading, but I'll add a comment.

keraf replied:
cf1975a

goodhoko reviewed (approved):
Didn't manage to go trhu this all. Posting what I got. I'm not sure I can review this in its entirety. .D Let's cross our fingers and rely on QA.

tomasklim commented on ci/packages/suite-web.yml:
```diff
@@ -81,7 +81,7 @@ suite-web deploy dev:
     COMPOSE_PROJECT_NAME: $CI_JOB_ID
     COMPOSE_FILE: ./docker/docker-compose.suite-ci.yml
-    # corresponds with the NextJs assetPrefix. Tells Cypress where is the index of application
+    # corresponds with the NextJs ASSET_PREFIX. Tells Cypress where is the index of application
```
```suggestion
    # corresponds with the environment variable ASSET_PREFIX defined in webpack config. Tells Cypress where is the index of application
```

keraf replied:
cf1975a

tomasklim commented on docs/misc/build.md:
```diff
@@ -0,0 +1,21 @@
+The front-end build of Suite is handled by Webpack configurations inside the `suite-build` package.
+
+The folder structure is as follows:
+- configs: Contains the Webpack configuration files. `base.webpack.config.ts` serves as a comon base for all other configurations. The other files in this folder are project specific such as `web.webpack.config.ts` or `desktop.webpack.config.ts` for `suite-web` and `suite-desktop` respectively. 
```
```suggestion
- configs: Contains the Webpack configuration files. `base.webpack.config.ts` serves as a common base for all other configurations. The other files in this folder are project specific such as `web.webpack.config.ts` or `desktop.webpack.config.ts` for `suite-web` and `suite-desktop` respectively. 
```

keraf replied:
cf1975a

tomasklim commented on packages/suite-build/configs/base.webpack.config.ts:
```diff
@@ -0,0 +1,148 @@
+        alias,
+        fallback: {
+            // Polyfills API for NodeJS libraries in the browser
+            crypto: require.resolve('crypto-browserify'),
```
Not sure how exactly fallback works here, but do we need to polyfill this library (77kB)? As window.crypto is built in. 

Same applies for the other libraries, probably we do not need to have e.g. 'path' on client side. I know that you did not do that, just a question.

szymonlesisz replied:
i'm not sure about crypto/stream. but `path` can be omitted. i did it in connect like this and it's working:
https://github.com/trezor/connect/blob/f309fc50060db6623ef1226b6204ea3c387e7877/webpack/config.dev.babel.js#L104

szymonlesisz replied:
check [this](https://www.mixmax.com/engineering/requiring-node-builtins-with-webpack/) regarding crypto
 ```
 module.exports = {
  externals: {
    'crypto': 'crypto'
  }
};
 ```

keraf replied:
16116f3

tomasklim commented on packages/suite-build/configs/desktop.webpack.config.ts:
```diff
@@ -0,0 +1,73 @@
+            filename: path.join(baseDir, 'build', 'index.html'),
+        }),
+        ...(isDev
+            ? [new webpack.HotModuleReplacementPlugin(), new ReactRefreshWebpackPlugin()]
```
is HotModuleReplacementPlugin still required with ReactRefreshWebpackPlugin?

react-refresh-webpack-plugin says 
> If you use webpack-dev-server or webpack-plugin-serve, you can set devServer.hot/devServer.hotOnly and hmr to true respectively, instead of adding the HMR plugin to your plugin list.

keraf replied:
fc60cbc

tomasklim commented on packages/suite-build/configs/web.webpack.config.ts:
```diff
@@ -0,0 +1,138 @@
+        ],
+    },
+    plugins: [
+        new webpack.DefinePlugin({
```
There are duplicated envs in suite-web and suite-desktop webpack configs. What about extracting duplicated ones to a new file?

keraf replied:
fc60cbc

tomasklim commented on packages/suite-build/configs/web.webpack.config.ts:
```diff
@@ -0,0 +1,138 @@
+              ]
+            : []),
+        ...(isDev
+            ? [new webpack.HotModuleReplacementPlugin(), new ReactRefreshWebpackPlugin()]
```
HotModuleReplacementPlugin => Same as for desktop

keraf replied:
fc60cbc

tomasklim commented on packages/suite-build/package.json:
```diff
@@ -0,0 +1,54 @@
+    "description": "trezor suite build",
+    "scripts": {
+        "dev:web": "TS_NODE_PROJECT=\"tsconfig.json\" webpack serve --config ./configs/web.webpack.config.ts --progress",
+        "build:web": "NODE_ENV=production TS_NODE_PROJECT=\"tsconfig.json\" webpack --config ./configs/web.webpack.config.ts --progress",
```
Note: During message system feature development I was a little bit confused that builds on suite.corp.sldev have production NODE_ENV and it is a production build.

keraf replied:
We want QA to test production builds on sldev. Maybe in the future we can make environment specific builds but as of now this is correct.

tomasklim commented on packages/suite-web/package.json:
```diff
@@ -6,33 +6,30 @@
-        "copy-files": "yarn workspace @trezor/suite-data copy-static-files",
-        "dev": "yarn workspace @trezor/suite-data copy-static-files && next",
+        "dev": "rimraf ./build && yarn workspace @trezor/suite-build run dev:web",
+        "dev:analyze": "ANALYZE=true yarn dev",
```
What about analyze prod? I think it is more important than dev

Btw is it even possible to run prod version on localhost easily?

keraf replied:
I renamed the command to `analyze` and it's now running a prod build (2b12f03).

Not really easily, but I will be looking at introducing `.env` files in the second part of this change (#3670).

tomasklim commented on packages/suite-build/utils/suite.ts:
```diff
@@ -0,0 +1,12 @@
+    });
+
+    return groups;
+};
```
What should this do? According to bundle analyzer all parts are still in one chunk. 

I would maybe try to split the vendor bundle more as it is still quite huge in comparison to the others.

keraf replied:
This was removed. Vendor bundle will be split more in the future. For now this is better than what we had.

tomasklim commented on packages/suite-data/package.json:
```diff
@@ -9,24 +9,18 @@
         "test-same-svgs": "ts-node ./test/test-same-svgs.ts"
     },
-    "dependencies": {
-        "ua-parser-js": "^0.7.24"
```
I agree with šzymon, that moving browser detection to project is not a good idea as we want to inform user even in browsers which are older than:
Chrome >= 59
Firefox >= 54

I think that if someone opens suite in IE11 or even chrome <39, it will not load at all.

IMHO best solution is to detect browser on server which serves the suite application. (I read somewhere that e.g. with nginx it is possible, not sure how to do it). This will result into saving some data and time to users who have new browsers.

szymonlesisz replied:
not sure if we are able to touch server configs.
suite-web is running on some amazon aws.
second thing is that it will require additional maintenance by admins (sldev) and it will not work on localhost

keraf replied:
Browser detect has been reverted (c82d5f6).

tomasklim commented on packages/suite-desktop/package.json:
```diff
@@ -7,13 +7,13 @@
     "homepage": "https://trezor.io/",
     "main": "dist/app.js",
     "scripts": {
-        "dev": "yarn dev:prepare && cross-env LAUNCH_ELECTRON=true node scripts/dev.js",
```
I think that cross-env is required to be able to run a project on Windows?

tomasklim commented on packages/suite-web/package.json:
```diff
@@ -6,33 +6,30 @@
-        "ts-loader": "^8.0.14"
+        "@types/react-router": "^5.1.11",
+        "@types/react-router-dom": "^5.1.7",
+        "@types/ua-parser-js": "^0.7.35"
```
How do we get ua-parser-js library? I do not see it in dependencies

keraf replied:
cf1975a

tomasklim commented on packages/suite-web/src/static/manifest.json:
```diff
@@ -0,0 +1,18 @@
+{
```
Side note: I think correct naming of this file is now site.webmanifest

keraf replied:
`manifest.json` is still valid but there's a dedicated extension for it now. I will update it to `.webmanifest`.

keraf replied:
f2775a8

tomasklim commented on packages/suite-web/src/support/BrowserDetect/constants.ts:
```diff
@@ -0,0 +1,29 @@
+export const supportedBrowsers = [
+    {
+        name: 'Chrome',
+        version: 69,
```
Versions in this file do not correspond with versions in `browserslist` file and `supportedBrowsers.ts`. 
Moreover update and download URLs are duplicite with `supportedBrowsers.ts`

keraf replied:
Browser detect has been reverted (c82d5f6). This is out of scope for this PR.

szymonlesisz reviewed (dismissed):
i've tried it locally and it is really few times faster than the previous configuration :) good job!
looking forward to use it in development (i'm going to do my work on top of this branch from now on)

yet my biggest concern is moving BrowserDetection in to app itself (commented below)

tomasklim commented on packages/suite/package.json:
```diff
@@ -29,6 +29,7 @@
         "file-saver": "^2.0.5",
         "framer-motion": "^3.2.1",
         "google-auth-library": "6.0.6",
+        "history": "^4.10.1",
```
History in version 4 have at least 2 bugs.

First one is that if you click on one link more times, it is stored in browser history everytime. So if you click on Dashboard 5x, then if you hit 5x times back button, you are still on dashboard. 
https://github.com/ReactTraining/history/pull/570#issuecomment-476511221

Another one is something with encoding strings into URL (for example for filtering), but I think we do not use that. If yes, I can try to find out what was the exact problem.

szymonlesisz replied:
first issue could be resolved by adding condition to `routerActions.goto` action, something like:
`if (currentRoute === requestedRoute) return;`

keraf replied:
cc516a6

tomasklim reviewed (commented):
Working for me.
A lot of small things I found during the message system development is changed or fixed. Good job!

Few notes or questions below.

goodhoko commented on package.json:
```diff
@@ -46,6 +46,7 @@
         "@babel/preset-react": "^7.13.13",
```
GitHub [won't let me](https://github.com/isaacs/github/issues/1655) comment on an arbitrary line. This comment relates to the lines 26-30 in this file, concretely all the `suite:` prefixed scripts in the root `package.json`: Shouldn't these be either removed or updated to use the new scripts from the `suite-build` package?

goodhoko commented on packages/suite-build/configs/base.webpack.config.ts:
```diff
@@ -0,0 +1,148 @@
+        },
+    },
+    optimization: {
+        splitChunks: {
+            cacheGroups: {
+                react: {
+                    chunks: 'initial',
+                    name: 'react',
+                    test: /[\\/]node_modules[\\/]react/,
+                },
+                vendors: {
+                    chunks: 'initial',
+                    name: 'vendors',
+                    test: /[\\/]node_modules[\\/](?!react)/,
+                },
+                components: {
+                    chunks: 'initial',
+                    name: 'components',
+                    test: /[\\/]packages[\\/]components[\\/]/,
+                },
+                ...suiteCacheGroups,
+            },
+        },
```
The webpack docs [discourage](https://webpack.js.org/plugins/split-chunks-plugin/#configuration) custom `splitChunks` config. Could we maybe add a comment explaining why we're having this custom config?

keraf replied:
It's common practice. This is solving the issue of having too many or too heavy chunks.

goodhoko commented on packages/suite-build/utils/env.ts:
```diff
@@ -0,0 +1,13 @@
+const { NODE_ENV, ANALYZE, LAUNCH_ELECTRON, ASSET_PREFIX } = process.env;
+
+const ALLOWED = ['development', 'production'] as const;
+const DEFAULT = ALLOWED[0]; // development
+
+const env = NODE_ENV as typeof ALLOWED[number];
+const environment = NODE_ENV !== undefined && ALLOWED.includes(env) ? env : DEFAULT;
```
Could be simplified to
```suggestion
const environment = NODE_ENV === 'production' ? 'production' : 'development';
```

Preserving both the semantics and types.

keraf replied:
cf1975a

szymonlesisz commented:
additional UI remarks:
- first loader is not centered like it used to be
- error pages (loading too long, unsupported views etc.) are not centered as well (i guess it's some shared container css)
- chunk preloader should preload images (?) there is visible glitch/jump (example: going from settings to settings/coins)
- maybe we could ask Ondra for some nicer preloader for chunks (thin line, like discovery? idk)
- index.html is missing meta tags (see production build)

-----
- hot reload on reducers ends with error (example, run dev and change suiteReducer `initialState`, say "tor: true", doesn't really matter, and observe), i think previous setup had it disabled for this reason

goodhoko commented on packages/suite-web/src/static/index.html:
```diff
@@ -0,0 +1,30 @@
+        <noscript>This website requires JavaScript to run.</noscript>
+        <div id="app"></div>
+    </body>
+</html>
```
NL pls. .)

keraf replied:
cf1975a

goodhoko commented on packages/suite-web/src/static/manifest.json:
```diff
@@ -0,0 +1,18 @@
+    "description": "New app for Trezor hardware wallets.",
+    "theme_color": "#212121",
+    "background_color": "#e3e3e3"
+  }
```
NL pls .)

keraf replied:
cf1975a

goodhoko commented on packages/suite-web/src/support/BrowserDetect/index.tsx:
```diff
@@ -0,0 +1,46 @@
+}
+
+const BrowserDetect = ({ children }: Props) => {
+    const ua = useMemo(() => new UAParser().getResult(), []);
```
I'm no expert  but isn't
> Don't use user agent, use feature detection

a [widely spread principle](https://developer.mozilla.org/en-US/docs/Web/HTTP/Browser_detection_using_the_user_agent)?

tomasklim replied:
Not maintainable here, as you would have to check hundreds of features

goodhoko replied:
I feared so. .)

keraf commented on packages/integration-tests/projects/suite-web/run_tests.js:
```diff
@@ -121,7 +121,7 @@ async function runTests() {
             chromeWebSecurity: false,
             trashAssetsBeforeRuns: false,
-            defaultCommandTimeout: 15000,
+            defaultCommandTimeout: 20000,
```
Check if still needed.

keraf commented on packages/suite-build/utils/alias.ts:
```diff
@@ -0,0 +1,24 @@
+import path from 'path';
+import { compilerOptions } from '../../../tsconfig.json';
```
Source of truth for aliases is in root tsconfig file (should be in docs).

keraf replied:
cf1975a

keraf commented on packages/suite-desktop/src-electron/tsconfig.json:
```diff
@@ -25,6 +25,7 @@
         "noImplicitAny": true
     },
     "include": [
+        "index.d.ts",
```
Double check

keraf commented on packages/suite-desktop/src/global.d.ts:
```diff
@@ -0,0 +1,9 @@
+    export = value;
+}
+
+declare module '*.css';
```
Double check if needed (css in desktop?)

keraf replied:
cf1975a

keraf commented on .eslintrc.js:
```diff
@@ -101,5 +102,6 @@ module.exports = {
         'no-use-before-define': 'off',
         '@typescript-eslint/no-use-before-define': ['error'],
         'require-await': ['error'],
+        '@typescript-eslint/triple-slash-reference': 'off',
```
Maybe better to add an exception for just the global.d.ts file instead of disabling the whole rule.

keraf replied:
cf1975a

keraf commented on packages/suite/global.d.ts:
```diff
@@ -46,5 +46,8 @@ declare global {
         desktopApi?: DesktopApi; // Electron API
+        Cypress?: any;
+        TrezorConnect?: any;
+        store?: any; // Cypress
```
Document/Comment

keraf commented on packages/suite/src/constants/suite/routes.ts:
```diff
@@ -1,223 +1,8 @@
 import { ArrayElement, ConstWithOptionalFields } from '@suite/types/utils';
```
This file only has types in it, so maybe route/types?

keraf replied:
Types are based of route config and only used there. I think it's fine to leave this as is.

keraf commented on packages/suite/src/workers/typings/index.d.ts:
```diff
@@ -1,5 +1,5 @@
-// https://github.com/zeit/next.js/issues/4768
-declare module 'worker-loader?name=static/[hash].worker.js!*' {
+// https://webpack.js.org/loaders/worker-loader/#integrating-with-typescript
+declare module 'worker-loader!*' {
```
Still needed?

keraf replied:
cf1975a

keraf commented:
Latest fixup (fc60cbc) contains a few more significant changes (to address the comments in this PR):
- Splitting out dev config in separate file.
- Single webpack config entry point, project is set via environment variable. 
- Reduced configuration re-use in web and desktop configs.
- Moved to webpack-serve (from webpack-dev-server). The local dev server is using the `ramdisk` feature which should speed up builds. If this causes issues, I will add another environment variable to disable it.
- Switched out webpack-cli with webpack-nano (smaller and simpler dependency).
- Changed Browser Detect to now only load react if needed. It is not a React app anymore (this will be extracted into a shared package once connect is using it).

This commit should also address the HMR refresh issue @szymonlesisz mentioned. If react-refresh cannot correctly re-render something, it will refresh the page.

This change also contains a patch for `@pmmmwh/react-refresh-webpack-plugin`. A PR has been opened in the project to address this: pmmmwh/react-refresh-webpack-plugin#374. The patch will be removed once/if the change gets merged and published.

szymonlesisz commented:
![Screenshot from 2021-05-18 12-13-17](https://user-images.githubusercontent.com/3435913/118634177-8b1ed180-b7d2-11eb-9bd9-3aabdc5e31c3.png)

is it me or the address (localhost) is missing?

keraf commented:
> ![Screenshot from 2021-05-18 12-13-17](https://user-images.githubusercontent.com/3435913/118634177-8b1ed180-b7d2-11eb-9bd9-3aabdc5e31c3.png)
> 
> is it me or the address (localhost) is missing?

That's expected. Any hostname pointing to this port will serve the dev site. I think there should be a parameter to manually set the hostname but imo it's fine like this.

szymonlesisz commented:
> That's expected. Any hostname pointing to this port will serve the dev site. I think there should be a parameter to manually set the hostname but imo it's fine like this.

i was used to click on link which appeared in console after successful compilation, now this is no longer an option

tomexx commented:
Just FYI. I recently added `raw-loader` to both suite-web and suite-desktop to be able to load markdown files stored in suite-data (for Suite Guide/Learn and Discover section).

matejkriz commented:
I'm on the rebase...

matejkriz commented on packages/suite-web/src/support/Router.tsx:
```diff
@@ -0,0 +1,60 @@
+        () => import('@wallet-views/coinmarket/exchange/offers'),
+    ),
+    'wallet-coinmarket-spend': lazy(() => import('@wallet-views/coinmarket/spend')),
+    'wallet-coinmarket-redirect': lazy(() => import('@wallet-views/coinmarket/redirect')),
```
I think this will not work because it's not defined in `routes` so it will have no defined Route in `routes.map`. Same in suite-desktop...

matejkriz replied:
fix: https://github.com/trezor/trezor-suite/pull/3446/commits/f7daa6cdfced404fa909a0b3b297c04e3ab763e6

tomasklim commented:
Before merging, please review and edit row 67 in `message_system.md`. Thank you

matejkriz commented on packages/suite/src/config/suite/routes.ts:
```diff
@@ -121,12 +121,6 @@ const routes = [
     },
-    {
-        name: 'coinmarket-buy-detail',
-        pattern: 'accounts/coinmarket/buy/detail',
```
duplicate definition of `wallet-coinmarket-buy-detail`

matejkriz commented:
We should implement some possibility to dump redux state other than `__NEXT_REDUX_STORE__.getState()` as it is no more working without Next. It's used by QA - see comment https://github.com/trezor/trezor-suite/pull/3720#issuecomment-843629000

tomasklim commented on package.json:
```diff
@@ -68,7 +70,11 @@
+        "ts-node": "^9.1.1",
+        "tsconfig-paths": "^3.9.0",
+        "typescript": "4.1.3",
+        "webpack": "^5.28.0",
+        "webpack-nano": "^1.1.1"
```
Check if we need these deps here

tomasklim replied:
unable to compile the project even when `webpack` and `webpack-nano` is available in specific package folder

tomasklim commented on packages/components/styled.d.ts:
```diff
@@ -6,3 +6,8 @@ declare module 'styled-components' {
     // eslint-disable-next-line @typescript-eslint/no-empty-interface
     export interface DefaultTheme extends SuiteThemeColors {}
 }
+
+declare module '*.png' {
+    const value: any;
+    export = value;
+}
```
Check

tomasklim replied:
A remnant from moving browser detection plugin

tomasklim commented on packages/suite-build/configs/base.webpack.config.ts:
```diff
@@ -0,0 +1,172 @@
+                ignore: ['static/connect'], // connect does not contain source maps for now
+                cleanArtifacts: true,
+            }),
+    ].filter(Boolean),
```
`...`

tomasklim commented on packages/suite-build/configs/web.webpack.config.ts:
```diff
@@ -0,0 +1,101 @@
+                ],
+            }),
+        isDev && new ReactRefreshWebpackPlugin(),
+    ].filter(Boolean), // Filters out disabled plugins
```
`...`

tomasklim commented on packages/suite-desktop/src/index.tsx:
```diff
@@ -1,76 +1,67 @@
+import AppRouter from './support/Router';
+import DesktopUpdater from './support/DesktopUpdater';
 
-const Tor = dynamic(() => import('@suite-support/Tor'), { ssr: false });
```
Check lazy

tomasklim commented on packages/suite-web/src/Main.tsx:
```diff
@@ -0,0 +1,75 @@
+import React, { useEffect } from 'react';
```
web x desktop 
Main.tsx x index.tsx

tomasklim commented on packages/suite-web/src/index.ts:
```diff
@@ -0,0 +1,23 @@
+// @ts-ignore - Export not needed
```
Find out if we need it

tomasklim commented on packages/suite-web/src/static/404.html:
```diff
@@ -0,0 +1,30 @@
+<!DOCTYPE html>
```
Is it used?

tomasklim commented on packages/suite/src/actions/suite/routerActions.ts:
```diff
@@ -66,48 +65,48 @@ export const onLocationChange = (url: string) => (dispatch: Dispatch, getState:
+export const closeModalApp = (preserveParams = true) => (dispatch: Dispatch) => {
     dispatch(suiteActions.lockRouter(false));
-    // const route = findRoute(Router.pathname + window.location.hash);
+    // const route = findRoute(Router.pathname + history.location.hash);
```
delete

tomasklim commented on packages/suite/src/components/suite/modals/QrScanner/index.tsx:
```diff
@@ -146,14 +146,16 @@ const QrScanner = ({ onCancel, decision }: Props) => {
-                        style={{ width: '100%', borderRadius: '3px' }}
-                        showViewFinder={false}
-                    />
+                    <Suspense fallback={<Loader size={64} />}>
```
Change to `BundleLoader

tomasklim commented on packages/suite/src/config/suite/features.ts:
```diff
@@ -12,6 +12,7 @@ export const FLAGS = {
     EXPORT_TRANSACTIONS: true, // Display export option on transactions
     SEARCH_TRANSACTIONS: true, // Display search option on transactions
     DESKTOP_AUTO_UPDATER: true, // Runs auto updater code on desktop
+    PWA: false, // Enable PWA support for suite-web
```
Check

tomasklim commented on packages/suite/src/constants/suite/routes.ts:
```diff
@@ -1,234 +1,14 @@
 
 type RouteKeys =
     | keyof ArrayElement<typeof routes>
+    | 'isModal'
```
Check

tomasklim commented on packages/suite/src/middlewares/onboarding/__tests__/onboardingMiddleware.test.ts:
```diff
@@ -106,6 +99,8 @@ describe('onboardingMiddleware', () => {
                         isFullscreenApp: true,
                         params: undefined,
+                        exact: undefined,
+                        isModal: undefined,
```
Check

tomasklim commented on packages/suite/src/middlewares/suite/__tests__/redirectMiddleware.test.ts:
```diff
@@ -128,6 +121,8 @@ describe('redirectMiddleware', () => {
                             isForegroundApp: undefined,
                             isFullscreenApp: undefined,
+                            exact: true,
+                            isModal: undefined,
```
Check

tomasklim commented on packages/suite/src/middlewares/suite/__tests__/suiteMiddleware.test.ts:
```diff
@@ -97,6 +89,8 @@ describe('suite middleware', () => {
                         isFullscreenApp: true,
                         params: undefined,
+                        exact: undefined,
+                        isModal: undefined,
```
Check

tomasklim commented on shell.nix:
```diff
@@ -21,15 +21,15 @@ in
       docker         # for e2e tests running on localhost
       docker-compose # for e2e tests running on localhost
-      nodejs
+      nodejs-14_x
```
check with @vdovhanych @marekrjpolak

tomasklim commented:
- check page headers 🟢 
- 404 page  🟢  (when route does not exists, 404 is shown, has to be configured in nginx, does not work on localhost)
- check shell-exec 🟢  (created issue)
- test PWA 🟢   (removed)
- document suite-web/src/index  🟢   
- check Images preloader - images flickering with disabled cache and slow network 🟢   (preloader is back)
- check Trade feature on Desktop (works fine on web) 🟢   
- fix discovery error 🟢  (do not optimize connect by suite webpack, it fixed tests)
- do not bundle all suite-data 🟢

TODO @matejkriz:

- check Bundle Loader 🔴  (for some reason it works only on localhost)
  - probably we can delete it and do not load routes lazily
  - or we can load them lazily and make bundles of 4 categories - settings, index, coinmarket, wallet
  - I would prefer to make one bundle of routes and delete BundleLoader (because the preloader of images is back)
- update sentry condition to be available only if sentry auth token is set (in base webpack config)

matejkriz commented:
> * check page headers 🟢
> * 404 page  🟢  (when route does not exists, 404 is shown, has to be configured in nginx, does not work on localhost)
> * check shell-exec 🟢  (created issue)
> * test PWA 🟢   (removed)
> * document suite-web/src/index  🟢
> * check Images preloader - images flickering with disabled cache and slow network 🟢   (preloader is back)
> * check Trade feature on Desktop (works fine on web) 🟢
> * fix discovery error 🟢  (do not optimize connect by suite webpack, it fixed tests)
> * do not bundle all suite-data 🟢
> 
> TODO @matejkriz:
> 
> * check Bundle Loader 🔴  (for some reason it works only on localhost)
>   
>   * probably we can delete it and do not load routes lazily
>   * or we can load them lazily and make bundles of 4 categories - settings, index, coinmarket, wallet
>   * I would prefer to make one bundle of routes and delete BundleLoader (because the preloader of images is back)
> * update sentry condition to be available only if sentry auth token is set (in base webpack config)

🟢 Bundle loader is actually working even deployed version. But you have to only clear the cache before testing, not disable it completely (otherwise the spinner.svg is loading every time and it's almost the same size as most of the page chunks...)
🟢 sentry condition updated

goodhoko commented:
🎉🎉💦💦🖤

sorooris commented:
hey guys, does this imply there's going to be another staging build?

matejkriz commented:
@sorooris definitely not for this one, this was merged after freeze by intention to keep enough time to find any issues till next release...

sorooris commented:
> @sorooris definitely not for this one, this was merged after freeze by intention to keep enough time to find any issues till next release...

oh yeah just noticed that it's the former feat/tschuss-next branch that's been tested separately - may i move it to the next month's project?

matejkriz commented:
@sorooris yes please, move it, thanks!

bosomt commented:
is there something that can/need to be tested on our side ?
or we just test it as part of our complete regression test that we perform after freeze ?

matejkriz commented:
It has been already tested separately as @sorooris mentioned – nothing more to test right now, I believe.

bosomt commented:
@matejkriz sorry, i haven't read previous comments, i see that it was already tested as tschuss-next branch :)
